jobs/integration/test_opa.py

```diff
@@ -0,0 +1,439 @@
+
+
+@pytest.mark.clouds(["vsphere"])
+class TestOPAWebhook(OPATestBase):
+    async def test_opa_webhook_ready(self, opa_controller_manager):
+        await self.wait_for_units(opa_controller_manager)
+
+        log("Waiting for gatekeeper charm to be ready")
+        await wait_for_application_status(
+            self.k8s_model, "gatekeeper-controller-manager", status="active"
+        )
+
+    async def test_create_ns_no_policy(self, opa_controller_manager):
+        try:
+            await self.validate_create_ns()
+        except JujuRunError as e:
+            if e.output.startswith(
+                "Error from server (InternalError): Internal error occurred"
+            ):
+                # Try again as it might be transient/initialization related
+                await asyncio.sleep(5)
+                await self.validate_create_ns()
+            else:
+                raise
+
+    async def test_policy_enforcement(self, opa_controller_manager):
+        log("Creating policy and constraint crds")
+        await self.deploy_example_policy()
+        log("Test that the policy is enforced")
+        try:
+            await self.validate_create_ns_fail()
+            await self.validate_create_ns_with_label()
+        finally:
+            await self.destroy_example_policy()
+
+    async def test_audit_deploy(
+        self, opa_controller_manager, update_status_interval, storage_pool
+    ):
+        log("Deploying the audit charm")
+        audit = await self.k8s_model.deploy(
+            "gatekeeper-audit",
+            channel=self.tools.charm_channel,
+            trust=True,
+            storage={"audit-volume": {"pool": storage_pool}},
+            config={"audit-interval": 1},
+        )
+
+        log("Waiting for audit charm to be ready")
+        try:
+            await self.wait_for_units(audit)
+
+            await self.deploy_example_policy()
+            log("Test that the policy is enforced")
+            await self.validate_create_ns_fail()
+            await self.validate_create_ns_with_label()
+            await self._validate_audit_actions(audit.units[0])
+            await self.destroy_example_policy()
+            log("Validate that the policy is no longer enforced")
+            await self.validate_create_ns()
+        finally:
+            log("Deleting the audit charm")
+            await self.tools.run(
+                "juju",
+                "remove-application",
+                "-m",
+                self.tools.k8s_connection,
+                "--force",
+                "--destroy-storage",
+                "gatekeeper-audit",
+            )
+            await self.wait_for_app_removed("gatekeeper-audit")
+
+    async def test_manager_reconcile(
+        self, opa_controller_manager, update_status_interval
+    ):
+        await self.wait_for_units(
+            opa_controller_manager, error_status=("error",), expected_status="blocked"
+        )
```

Comment:
ahh -- looking for blocked units.  okay now i see the value of `wait_for_units`.